### PR TITLE
fix(iam): grantAssumeRole silently fails with service and account principals

### PR DESCRIPTION
**Pull request by kishiel**
### Reason for this change

grantAssumeRole silently fails if a Service Principal or Account Principal is used which led me to a false assumption about the correctness of a role's permission scope

### Description of changes

This change will throw an error if a Service Principal is used.  I was unable to find a way to accomplish the same behavior for Account Principals.

Documentation was updated to help guide a user to the appropriate function usage for Service and Account Principals.

### Description of how you validated changes

* Added a unit test
* This change required me to re-run two unrelated snapshot tests which were throwing errors outside of the scope of this change.

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-iam/README.md [hunk 1]
```diff
@@ -369,6 +369,37 @@ new iam.Role(this, 'Role', {
 });
 ```
 
+### Granting assume role permission from a role
+
+Principals can be granted permission to assume a role using `grantAssumeRole`.  Note that this does not apply to Service Principals or Account Principals as they must be added to the role Trust Policy.
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-iam/README.md:374]: Nit: lowercase `service principals` [1]
Nit: lowercase `account principals [2]

Also move the note to a newline. 

[1] https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_principal.html#principal-services
[2] https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_principal.html#principal-accounts

File: packages/aws-cdk-lib/aws-iam/README.md [hunk 2]
```diff
@@ -369,6 +369,37 @@ new iam.Role(this, 'Role', {
 });
 ```
 
+### Granting assume role permission from a role
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-iam/README.md:372]: "From a role" is confusing here. Can you change this to "Granting principals permission to assume a role"

File: packages/aws-cdk-lib/aws-iam/README.md [hunk 3]
```diff
@@ -369,6 +369,37 @@ new iam.Role(this, 'Role', {
...
+```ts
+const user = new iam.User(this, 'user')
+const role = new iam.Role(this, 'role', {
+  assumedBy: new iam.AccountPrincipal(this.account)
+});
+
+role.grantAssumeRole(user);
+```
+
+### Giving Service Principals and Account Principals assume role permission from a role
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-iam/README.md:385]: "Granting service and account principals permission to assume a role"

File: packages/aws-cdk-lib/aws-iam/README.md [hunk 4]
```diff
@@ -369,6 +369,37 @@ new iam.Role(this, 'Role', {
...
+
+Service Principals and Account Principals can be granted permission to assume a role using `assumeRolePolicy` which modifies the role Trust Policy.
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-iam/README.md:387]: lowercase

Comment (msambol) [reply packages/aws-cdk-lib/aws-iam/README.md:387]: `Trust Policy` should be lowercase too.

File: packages/aws-cdk-lib/aws-iam/lib/role.ts [hunk 5]
```diff
@@ -594,6 +594,10 @@ export class Role extends Resource implements IRole {
    * Grant permissions to the given principal to assume this role.
    */
   public grantAssumeRole(identity: IPrincipal) {
+    // Service Principals must use assumeRolePolicy
+    if (identity.policyFragment.principalJson.Service) {
+      throw new Error('Cannot use a Service Principal with grantAssumeRole, use assumeRolePolicy instead.');
+    }
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-iam/lib/role.ts:600]: I think this would work?
```
    if (identity instanceof ServicePrincipal || identity instanceof AccountPrincipal) {
      throw new Error('Cannot use a service or account principal with grantAssumeRole, use assumeRolePolicy instead.');
    }
```

**Comment by msambol** [comment]
Can you change the title? It should reflect the bug. Maybe:

`fix(iam): grantAssumeRole silently fails with service and account principals`

**Review by msambol — CHANGES_REQUESTED** [review]
See inline. Great start!

File: packages/aws-cdk-lib/aws-iam/lib/role.ts (hunk 5 above)
Comment (kishiel) [reply packages/aws-cdk-lib/aws-iam/lib/role.ts:600]: Good call, this appears to have worked.  Added another test for account principal.

**Comment by kishiel** [comment]
Incorporated all feedback and updated title

File: packages/aws-cdk-lib/aws-iam/README.md [hunk 6]
```diff
@@ -369,6 +369,39 @@ new iam.Role(this, 'Role', {
 });
 ```
 
+### Granting an identity permission to assume a role
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-iam/README.md:372]: Can you change this to "Granting a principal...". That's the more common verbiage.

File: packages/aws-cdk-lib/aws-iam/README.md [hunk 7]
```diff
@@ -369,6 +369,39 @@ new iam.Role(this, 'Role', {
...
+
+An identity can be granted permission to assume a role using `grantAssumeRole`.
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-iam/README.md:374]: Likewise here, "a principal."

**Review by msambol — COMMENTED** [review]
Few more comments then we're good to go.

File: packages/aws-cdk-lib/aws-iam/README.md (hunk 6 above)
Comment (kishiel) [reply packages/aws-cdk-lib/aws-iam/README.md:372]: Ah, I forgot I'd modified this.  I changed this to identity based on [this section](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-iam/README.md#principals-vs-identities) of the readme but I think I'm just misreading it.  Will change

**Review by msambol — APPROVED** [review]
Looks good. Thanks!

File: packages/@aws-cdk-testing/framework-integ/test/aws-iam/test/integ.managed-policy.ts [hunk 8]
```diff
@@ -24,7 +24,8 @@ const policy3 = ManagedPolicy.fromAwsManagedPolicyName('SecurityAudit');
 user.addManagedPolicy(policy3);
 
 const role = new Role(stack, 'Role', { assumedBy: new AccountRootPrincipal() });
-role.grantAssumeRole(policy.grantPrincipal);
+role.grantAssumeRole(user);
```
Comment (GavinZZ) [inline packages/@aws-cdk-testing/framework-integ/test/aws-iam/test/integ.managed-policy.ts:27]: Why is this change needed in the test file?

File: packages/aws-cdk-lib/aws-iam/README.md [hunk 9]
```diff
@@ -369,6 +369,39 @@ new iam.Role(this, 'Role', {
 });
 ```
 
+### Granting a principal permission to assume a role
```
Comment (GavinZZ) [inline packages/aws-cdk-lib/aws-iam/README.md:372]: Thanks for this clear documentation update.

**Review by GavinZZ — COMMENTED** [review]
Generally LGTM, just a question for clarification before approving.

File: packages/@aws-cdk-testing/framework-integ/test/aws-iam/test/integ.managed-policy.ts (hunk 8 above)
Comment (msambol) [reply packages/@aws-cdk-testing/framework-integ/test/aws-iam/test/integ.managed-policy.ts:27]: I think to prove you can still use `grantAssumeRole` on an IAM user.

Comment (kishiel) [reply packages/@aws-cdk-testing/framework-integ/test/aws-iam/test/integ.managed-policy.ts:27]: This was a preexisting fault that bubbled up when I made this change.  The error was something to the effect that you cannot grant a policy assume permission.  Since the apparent intent of this portion was to prove that grantAssume could be given to the policy's principal (the user) I changed it to that.

Comment (kishiel) [reply packages/@aws-cdk-testing/framework-integ/test/aws-iam/test/integ.managed-policy.ts:27]: I'm taking another look at this to see if I can reproduce it

Comment (kishiel) [reply packages/@aws-cdk-testing/framework-integ/test/aws-iam/test/integ.managed-policy.ts:27]: I can't get this to reproduce.  I've reverted the change and re-run the snapshot.  Strange.  Pushing an update.

**Review by GavinZZ — DISMISSED** [review]

**Review by GavinZZ — APPROVED** [review]
